File: src/main/scala/examples/DeathWatchActor.scala
package examples

import akka.actor.{Actor, ActorSystem, Props, Terminated}

class DeathWatchActor extends Actor {

  override def preStart() {
    val volatileGreetingActor = context.actorOf(Props[VolatileGreetingActor])
    context.watch(volatileGreetingActor)
    volatileGreetingActor ! "print this message, please!"
  }

  def receive = {
    case Terminated(_) => println("looks like an actor has died :(")
  }
}

class VolatileGreetingActor extends Actor {

  def receive = {
    case s: String => {
      println("stopping")
      context.stop(self)
    }
  }
}

object DeathWatchActor extends App {
  val system = ActorSystem("MySystem")
  system.actorOf(Props[DeathWatchActor])

  system.shutdown()
}

File: src/main/java/examples/DeathWatchActor.java
package examples;

import akka.actor.*;
import akka.japi.pf.ReceiveBuilder;
import scala.PartialFunction;
import scala.runtime.BoxedUnit;

public class DeathWatchActor extends AbstractActor {

    @Override
    public void preStart() throws Exception {
        ActorRef greetingActor = context().actorOf(Props.create(VolatileGreetingActor.class), "volatileActor");
        context().watch(greetingActor);
        greetingActor.tell("print this message, please!", self());
    }

    @Override
    public PartialFunction<Object, BoxedUnit> receive() {
        return ReceiveBuilder.match(
                Terminated.class, s -> {
                    System.out.println("Looks like an actor has died");
                    context().stop(self());
                }
        ).build();
    }

    public static void main(String[] args) throws InterruptedException {
        ActorSystem system = ActorSystem.create("MySystem");
        system.actorOf(Props.create(DeathWatchActor.class), "deathWatchActor");

        // There are better ways to ensure message are received before termination
        Thread.sleep(100);
        system.shutdown();
    }
}
